firefox-ios/Client/Frontend/Browser/Toolbars/Redux/AddressBarState.swift

```diff
@@ -124,449 +124,504 @@ struct AddressBarState: StateType, Equatable {
-                                         addressBarState: state,
-                                         isEditing: state.isEditing,
-                                         showQRPageAction: false),
-                browserActions: state.browserActions,
-                borderPosition: state.borderPosition,
-                url: state.url,
-                searchTerm: state.searchTerm,
-                lockIconImageName: state.lockIconImageName,
-                isEditing: state.isEditing,
-                isScrollingDuringEdit: state.isScrollingDuringEdit,
-                shouldSelectSearchTerm: state.shouldSelectSearchTerm,
-                isLoading: state.isLoading,
-                readerModeState: state.readerModeState,
-                didStartTyping: true,
-                showQRPageAction: false
-            )
+            return handleDidEnterSearchTermAction(state: state, action: action)
 
         case ToolbarActionType.didStartTyping:
-            guard let toolbarAction = action as? ToolbarAction else { return state }
-
-            return AddressBarState(
-                windowUUID: state.windowUUID,
-                navigationActions: state.navigationActions,
-                pageActions: state.pageActions,
-                browserActions: state.browserActions,
-                borderPosition: state.borderPosition,
-                url: state.url,
-                searchTerm: state.searchTerm,
-                lockIconImageName: state.lockIconImageName,
-                isEditing: state.isEditing,
-                isScrollingDuringEdit: state.isScrollingDuringEdit,
-                shouldSelectSearchTerm: state.shouldSelectSearchTerm,
-                isLoading: state.isLoading,
-                readerModeState: state.readerModeState,
-                didStartTyping: true,
-                showQRPageAction: state.showQRPageAction
-            )
+            return handleDidStartTypingAction(state: state, action: action)
 
         default:
-            return AddressBarState(
-                windowUUID: state.windowUUID,
-                navigationActions: state.navigationActions,
-                pageActions: state.pageActions,
-                browserActions: state.browserActions,
-                borderPosition: state.borderPosition,
-                url: state.url,
-                searchTerm: state.searchTerm,
-                lockIconImageName: state.lockIconImageName,
-                safeListedURLImageName: state.safeListedURLImageName,
-                isEditing: state.isEditing,
-                isScrollingDuringEdit: state.isScrollingDuringEdit,
-                shouldSelectSearchTerm: state.shouldSelectSearchTerm,
-                isLoading: state.isLoading,
-                readerModeState: state.readerModeState,
-                didStartTyping: state.didStartTyping,
-                showQRPageAction: state.showQRPageAction
-            )
+            return state
```

Comment:
I noticed a similar comment [here](https://github.com/mozilla-mobile/firefox-ios/pull/22636#discussion_r1806647143) where @thatswinnie mentioned it. I also saw [your comment](https://github.com/mozilla-mobile/firefox-ios/pull/22679#discussion_r1821381913) in the another PR, and I believe we're talking about the same problem. I'm a bit concerned I may have done this in other PRs as well. It's fixed here, and I'll review other places where I made similar changes.

Reply:
I have an update: I found [this comment](https://github.com/mozilla-mobile/firefox-ios/pull/22502#discussion_r1796835692) in a previous PR, where I initially implemented the `default` case by returning a new instance. However, I received feedback suggesting that it may not be a problem to use the same instance instead.

My understanding is that the choice between returning the same instance or a new one might depend on how this object will be used. But creating a new instance of the struct using the same values should not be a problem because Swift structs follow a copy on write model. So we shouldn't have any issues for memory usage or performance here.

I would really appreciate any insights you might have on the best approach here. Once I fully understand this, I'll be able to apply the correct method to another PR I created.

Reply:
That's a really good point, thank you for bringing it up. I opened a [PR](https://github.com/mozilla-mobile/firefox-ios/pull/22812) to add a method to ScreenState to return a default State. if that is merged you can rebase and we can implement that default method directly in this state. For the default state we should only return the required property from the state initializer so we clear any transient state from the previous state.

Reply:
I think we can merge your implementation so far and leave my PR for later. Thank you again your work was great 🎉

Reply:
Okay, @FilippoZazzeroni. Your PR is related to another change https://github.com/mozilla-mobile/firefox-ios/pull/22502. We can move forward with these changes for now and then incorporate your implementation once your PR is merged. Thanks for the feedback and patience.

Reply:
No problem @ionixjunior. Yes i still have to finish that implementation. But anyway thanks again for the contributions 🎉 i'm going to merge this PR